smtpserver/server.go

```diff
@@ -1899,6 +1900,34 @@ func (c *conn) submit(ctx context.Context, recvHdrFor func(string) string, msgWr
 		xsmtpUserErrorf(smtp.C550MailboxUnavail, smtp.SePol7DeliveryUnauth1, "must match authenticated user")
 	}
 
+	// Check if the message contains non-ascii characters. If no such characters are found,
```

Comment:
i think we should do the smtputf8-check earlier. already in cmdData. just above the comment "Prepare "Received" header." (github PR review don't seem to let me make a comment there...). because we start composing headers based on c.smtputf8 from then on.

i think we should also keep a separate flag for whether we use smtputf8 for the message, vs what the incoming transaction has. perhaps we should add a flag "msgutf8" or "outsmtputf8" or something similar, and use that when composing headers and when adding to the queue. the smtputf8 flag is used when parsing commands. setting it to false after a the incoming connection has enabled it could lead to confusing/incorrect errors.

it would also be good to add a log line when we change from smtputf8 to non-utf8.

Reply:
Done.

About the separate `msgsmtputf8` flag, I must confess that I've replaced all occurrences of `c.smtputf8` found in server.go with the new flag without checking much (tests are still all ok, so we must be good...).

Reply:
yes, this is almost entirely fine. but the few calls to newParser should use c.smtputf8.

it's hard to trigger a problem around that: you would have to trigger an error in cmdData after setting msgsmtputf8 to false, then (as client) do another rcptto with a utf8 address. not even sure that would be a valid sequence of commands.